File: src/main/java/me/hhitt/disasters/model/block/DisappearBlock.java
package me.hhitt.disasters.model.block;

import me.hhitt.disasters.arena.Arena;
import me.hhitt.disasters.disaster.DisasterRegistry;
import org.bukkit.Location;
import org.bukkit.Material;

import java.util.Objects;

public class DisappearBlock {

    private static final int MAX_STAGE = 9;
    private static final int STAGE_INCREMENT = 2;

    private final Arena arena;
    private final Location location;
    private final int sourceId;
    private int currentStage = -1;
    private boolean occupied = true;

    public DisappearBlock(Arena arena, Location location) {
        this.arena = arena;
        this.location = location;
        this.sourceId = Objects.hash(
            location.getWorld() != null ? location.getWorld().getUID() : null,
            location.getBlockX(),
            location.getBlockY(),
            location.getBlockZ()
        );
    }

    public Location getLocation() {
        return location;
    }

    public boolean isOccupied() {
        return occupied;
    }

    public void setOccupied(boolean occupied) {
        this.occupied = occupied;
    }

    public void updateMaterial() {
        if (!occupied) {
            return;
        }
        currentStage += STAGE_INCREMENT;
        if (currentStage <= MAX_STAGE) {
            sendCrackAnimation(currentStage);
        } else {
            clearCrackAnimation();
            setBlockToAir();
            DisasterRegistry.removeBlockFromDisappear(arena, this);
        }
    }

    private void sendCrackAnimation(int stage) {
        float progress = Math.max(0.0F, Math.min(1.0F, (float) stage / (float) MAX_STAGE));
        arena.getPlaying().forEach(player -> player.sendBlockDamage(location, progress, sourceId));
    }

    private void clearCrackAnimation() {
        arena.getPlaying().forEach(player -> player.sendBlockDamage(location, 0.0F, sourceId));
    }

    private void setBlockToAir() {
        location.getBlock().setType(Material.AIR, false);
    }
}

File: src/main/kotlin/me/hhitt/disasters/model/block/DisappearBlock.kt
package me.hhitt.disasters.model.block

import me.hhitt.disasters.arena.Arena
import me.hhitt.disasters.disaster.DisasterRegistry
import net.minecraft.core.BlockPos
import net.minecraft.network.protocol.game.ClientboundBlockDestructionPacket
import net.minecraft.network.protocol.game.ClientboundBlockUpdatePacket
import net.minecraft.world.level.block.Blocks
import org.bukkit.Location
import org.bukkit.craftbukkit.CraftWorld
import org.bukkit.craftbukkit.entity.CraftPlayer

class DisappearBlock(private val arena: Arena, val location: Location) {

    // 10 crack stages (0-9), then block breaks to air
    // Advances 2 stages per pulse (1 pulse = 1 second) so full break takes ~5 seconds
    private val maxStage = 9
    private val stageIncrement = 2
    private var currentStage = -1
    // Whether a player is currently standing on this block (controls pause/resume)
    var occupied = true

    /**
     * Each entity ID must be unique per block destruction overlay.
     * Using the block's hash as a stable ID so the client knows which block to crack.
     */
    private val entityId = location.hashCode()

    fun updateMaterial() {
        // Only advance crack stages while a player is standing on the block
        if (!occupied) return

        currentStage += stageIncrement
        if (currentStage <= maxStage) {
            sendCrackAnimation(currentStage)
        } else {
            // Cracking complete — destroy the block
            clearCrackAnimation()
            setBlockToAir()
            DisasterRegistry.removeBlockFromDisappear(arena, this)
        }
    }

    /**
     * Sends the block crack overlay packet (same visual as when mining a block by hand).
     * Stage 0 = first cracks, stage 9 = about to break.
     */
    private fun sendCrackAnimation(stage: Int) {
        val blockPos = BlockPos(location.blockX, location.blockY, location.blockZ)
        val packet = ClientboundBlockDestructionPacket(entityId, blockPos, stage)
        arena.playing.forEach { player ->
            (player as CraftPlayer).handle.connection.send(packet)
        }
    }

    /**
     * Clears the crack overlay by sending stage -1 (removes the texture).
     */
    private fun clearCrackAnimation() {
        val blockPos = BlockPos(location.blockX, location.blockY, location.blockZ)
        val packet = ClientboundBlockDestructionPacket(entityId, blockPos, -1)
        arena.playing.forEach { player ->
            (player as CraftPlayer).handle.connection.send(packet)
        }
    }

    private fun setBlockToAir() {
        val worldServer = (location.world as CraftWorld).handle
        val blockPos = BlockPos(location.blockX, location.blockY, location.blockZ)
        worldServer.setBlockAndUpdate(blockPos, Blocks.AIR.defaultBlockState())
        val packet = ClientboundBlockUpdatePacket(worldServer, blockPos)
        arena.playing.forEach { player ->
            (player as CraftPlayer).handle.connection.send(packet)
        }
    }
}